crates/uv-publish/src/lib.rs

```diff
@@ -354,86 +355,127 @@ pub async fn check_trusted_publishing(
     }
 }
 
-/// Upload a file to a registry.
-///
-/// Returns `true` if the file was newly uploaded and `false` if it already existed.
-///
-/// Implements a custom retry flow since the request isn't cloneable.
-pub async fn upload(
+/// Upload a file to a registry, with retries.
+pub async fn upload_with_retry(
     file: &Path,
     raw_filename: &str,
     filename: &DistFilename,
     registry: &Url,
     client: &BaseClient,
-    retries: u32,
     username: Option<&str>,
     password: Option<&str>,
     check_url_client: Option<&CheckUrlClient<'_>>,
     reporter: Arc<impl Reporter>,
 ) -> Result<bool, PublishError> {
-    let form_metadata = form_metadata(file, filename)
-        .await
-        .map_err(|err| PublishError::PublishPrepare(file.to_path_buf(), Box::new(err)))?;
-
-    // Retry loop
-    let mut attempt = 0;
+    let mut n_past_retries = 0;
+    let start_time = SystemTime::now();
+    let retry_policy = client.retry_policy();
     loop {
-        let (request, idx) = build_request(
+        let result = upload(
             file,
             raw_filename,
             filename,
             registry,
             client,
             username,
             password,
-            &form_metadata,
+            check_url_client,
             reporter.clone(),
         )
+        .await;
+        if result
+            .as_ref()
+            .err()
+            .is_some_and(|err| is_extended_transient_error(err))
```

Comment:
This now uses `is_extended_transient_error` instead of `UvRetryableStrategy.handle(&result) == Some(Retryable::Transient)`. Wouldn't the latter already be called in the middle itself...?

Reply:
@konstin -- just confirming, do you agree that this is ok?

Reply:
It will miss the "default" cases handled by the retry strategy, like other transient request errors.

Reply:
I think that's okay... presuming that the default strategy doesn't cover anything that would occur at response read time.

Reply:
Thanks for the ping, i didn't realize we were missing that layer now. I don't have any reference upload intermittent errors, but I think we should preserve the `UvRetryableStrategy`, i can see things like timeouts and cancelled responses happening here too.

Reply:
Ok.

Reply:
So that we're on the same page: The retry middleware clones the request, and then resends it if it failed (https://github.com/TrueLayer/reqwest-middleware/blob/8a494c165734e24c62823714843e1c9347027e8a/reqwest-retry/src/middleware.rs#L150-L155). But a POST request with formdata and a streaming file isn't cloneable (limitation of reqwest, https://github.com/seanmonstar/reqwest/issues/2416). That's why the upload client is building a custom base client with `.retries(0)` that never internally tries to clone the request and instead has a loop that always rebuilds the request.

Reply:
Given that, it ended up being much simpler to leave it all in a single method.

Reply:
Ohh ok, I missed the `.retries(0)`.